src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -2101,12 +2176,25 @@ static int parse_args(vector<const char*>& args, std::ostream *err_msg,
     return -EINVAL;
   }
 
+  std::string cookie;
   switch (cmd) {
     case Attach:
       if (cfg->devpath.empty()) {
         *err_msg << "rbd-nbd: must specify device to attach";
         return -EINVAL;
       }
+      // Allowing attach without --cookie option for kernel without
+      // NBD_ATTR_BACKEND_IDENTIFIER support for compatibility
+      cookie = get_cookie(cfg->devpath);
+      if (!cookie.empty()) {
```

Comment:
I think this is wrong.  While `rbd-nbd attach` may allow attaching without `--cookie`, `rbd device attach` should not.
`rbd device attach` semantics should be:
- if no `--cookie`, require `--force` (already there)
- if `--cookie`, fail for kernels without the patch (seems to be missing)

Reply:
I don't think map displays the cookie even if `--show-cookie` is mentioned when the kernel changes are missing. So wondering from where do they even supply the cookie from at the time of attach?

If it is just to defend in the code, I can add one.

Reply:
It is to protect the (sometimes clueless) user from shooting themselves in the foot.  Consider:
- argh, the command without `--cookie` doesn't work, wants `--force`
- but `--force` sounds dangerous, especially in a script (non-interactive)...
- what if I pass `--cookie 123` instead?
- it works!

Also, in the future we might want to support user-supplied cookies, passed on map.  In that case the cookie could be stored somewhere (or they could choose to use the image name, etc).

Reply:
OK I hear that as an ack for user-supplied cookie :-)
I can add a quick change.